### ISSUE TITLE
Refactor multi_io_context

### ISSUE THREAD

**Issue by vinniefalco** [boostorg/beast#1784]
This class could use a rewrite and also I should probably live-stream the changes.

**Comment by benstadin** [comment]
+1 I‘m interested to further boost performance for multithreaded servers. This seems to be a bottle neck in my application as the connections per second do not scale beyond a certain limit, regardless of CPU cores available.

**Comment by vinniefalco** [comment]
> connections per second

Do you mean the number of incoming connections per second? If you want to maximize that number you need multiple listening sockets, all listening on the same port (and the appropriate socket options set). Note that I am by no means an expert at this.

**Comment by benstadin** [comment]
Yes, that‘s what I actually mean. I‘ve tried some code snippets I found in one of your repos (though as I remember those snippets were from somebody else), but gave up on it since it required some deeper insights into ASIO and Beast. I thought this is related.

**Comment by vinniefalco** [comment]
None of my examples address getting a high number of received connections, and I don't have any experience with it, except that in order to get high rates you need multiple listening sockets.

**Comment by benstadin** [comment]
I‘ve found our previous conversation [1]. If I understand correctly, for maximum throughput the best would be to use something like number of CPU cores for the multi IO contexts and each context has its own listening socket?

[1] https://github.com/boostorg/beast/issues/1285

**Comment by vinniefalco** [comment]
I'm not sure, you would have to experiment, You might even need multiple listening sockets for each `io_context`.

**Comment by AmitArie** [comment]
Hey guys!
There is something new about this subject?
The idea sounds very interesting!

I also want to find a way to utilize the amount of cores my server has, and to support massive amount of connections per second.

**Comment by madmongo1** [comment]
There are some videos of Chris Kohlhoff giving talks on just this subject our there on the internet.

**Comment by vinniefalco** [comment]
I have some code here
https://github.com/vinniefalco/beast/tree/multi-io/include/boost/beast/_experimental/core

**Comment by klemens-morgenstern** [comment]
Shuoldn't this just be done with `asio::thread_pool` nowadays?
